File: backend/app/services/youtube_service.py
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from typing import List, Dict, Any, Tuple
from app.config import settings


class YouTubeService:
    def __init__(self):
        self._client = None

    def _get_client(self):
        if self._client is None:
            self._client = build("youtube", "v3", developerKey=settings.youtube_api_key)
        return self._client

    def get_video_metadata(self, video_id: str) -> Dict[str, Any]:
        try:
            response = (
                self._get_client()
                .videos()
                .list(part="snippet,statistics", id=video_id)
                .execute()
            )
            if not response.get("items"):
                raise ValueError(f"Video {video_id} not found or is private")
            item = response["items"][0]
            snippet = item["snippet"]
            thumbnails = snippet.get("thumbnails", {})
            thumbnail = (
                thumbnails.get("maxres", {}).get("url")
                or thumbnails.get("high", {}).get("url")
                or thumbnails.get("medium", {}).get("url")
                or ""
            )
            return {
                "title": snippet.get("title", "Unknown Title"),
                "thumbnail": thumbnail,
                "channel": snippet.get("channelTitle", ""),
                "view_count": int(item.get("statistics", {}).get("viewCount", 0)),
                "comment_count": int(item.get("statistics", {}).get("commentCount", 0)),
            }
        except HttpError as e:
            raise RuntimeError(f"YouTube API error: {e.reason}")

    def fetch_comments(self, video_id: str) -> List[Dict[str, Any]]:
        comments = []
        next_page_token = None
        client = self._get_client()

        try:
            while len(comments) < settings.max_comments:
                kwargs: Dict[str, Any] = {
                    "part": "snippet",
                    "videoId": video_id,
                    "maxResults": 100,
                    "textFormat": "plainText",
                    "order": "relevance",
                }
                if next_page_token:
                    kwargs["pageToken"] = next_page_token

                response = client.commentThreads().list(**kwargs).execute()

                for item in response.get("items", []):
                    top = item["snippet"]["topLevelComment"]["snippet"]
                    comments.append(
                        {
                            "id": item["id"],
                            "text": top.get("textDisplay", ""),
                            "author": top.get("authorDisplayName", "Anonymous"),
                            "likes": top.get("likeCount", 0),
                            "published_at": top.get("publishedAt", ""),
                        }
                    )

                next_page_token = response.get("nextPageToken")
                if not next_page_token:
                    break

        except HttpError as e:
            if "commentsDisabled" in str(e):
                raise RuntimeError("Comments are disabled for this video")
            raise RuntimeError(f"YouTube API error: {e.reason}")

        return comments[: settings.max_comments]


youtube_service = YouTubeService()

File: backend/src/services/youtubeService.js
const { google } = require('googleapis');
const config = require('../config');

class YouTubeService {
  constructor() {
    this._client = null;
  }

  _getClient() {
    if (!this._client) {
      this._client = google.youtube({ version: 'v3', auth: config.youtubeApiKey });
    }
    return this._client;
  }

  async getVideoMetadata(videoId) {
    let response;
    try {
      response = await this._getClient().videos.list({
        part: ['snippet', 'statistics'],
        id: [videoId],
      });
    } catch (err) {
      throw new Error(`YouTube API error: ${err.message}`);
    }

    const items = response.data.items;
    if (!items || items.length === 0) {
      throw new Error(`Video ${videoId} not found or is private`);
    }

    const item = items[0];
    const snippet = item.snippet;
    const thumbnails = snippet.thumbnails || {};
    const thumbnail =
      thumbnails.maxres?.url ||
      thumbnails.high?.url ||
      thumbnails.medium?.url ||
      '';

    return {
      title: snippet.title || 'Unknown Title',
      thumbnail,
      channel: snippet.channelTitle || '',
      view_count: parseInt(item.statistics?.viewCount || 0),
      comment_count: parseInt(item.statistics?.commentCount || 0),
    };
  }

  async fetchComments(videoId) {
    const comments = [];
    let nextPageToken = null;
    const client = this._getClient();
    console.log(`[fetchComments] maxComments=${config.maxComments}`);

    try {
      while (comments.length < config.maxComments) {
        const params = {
          part: ['snippet'],
          videoId,
          maxResults: 100,
          textFormat: 'plainText',
          order: 'relevance',
        };
        if (nextPageToken) params.pageToken = nextPageToken;

        const response = await client.commentThreads.list(params);
        const items = response.data.items || [];

        for (const item of items) {
          const top = item.snippet.topLevelComment.snippet;
          comments.push({
            id: item.id,
            text: top.textDisplay || '',
            author: top.authorDisplayName || 'Anonymous',
            likes: top.likeCount || 0,
            published_at: top.publishedAt || '',
          });
        }

        nextPageToken = response.data.nextPageToken;
        if (!nextPageToken) break;
      }
    } catch (err) {
      if (err.message && err.message.includes('commentsDisabled')) {
        throw new Error('Comments are disabled for this video');
      }
      throw new Error(`YouTube API error: ${err.message}`);
    }

    return comments.slice(0, config.maxComments);
  }
}

module.exports = new YouTubeService();
